File: Scripts/Functions/simulation_summary_functions.R
###################################################################################
#### Penalized estimation in finite mixtures of multivariate regression models ####
####                         via the EM-PGM algorithm                          ####
###################################################################################
####                           Author: Heeyeon Kang                            ####
####                         Supervisor: Sunyoung Shin                         ####
###################################################################################
####         R-code for numerical presentation of simulation results.          ####
###################################################################################

# The following R-code includes functions to numerically present the results of the 
# simulations in Section 5.1 - 5.3.


### Packages ###
requiredPackages <- c("MASS", "pracma", "expm", "corrcoverage", 
                      "causact", "fossil", "flexmix", "gtools")
for(p in requiredPackages){
  if(!require(p, character.only=TRUE)) install.packages(p)
  library(p, character.only=TRUE)
}

TPR <- function(true, pred){
  true_Bk <- true$Bk
  pred_Bk <- pred$Bk
  
  P <- permutations(K, K)
  norm_of_true_and_pred_Bk <- vector()
  for(i in 1:nrow(P)){
    perm_Bk <- pred_Bk[P[i,]]
    norm_of_true_and_pred_Bk[i] <- sum(sapply(1:K, function(t) norm(true_Bk[[t]] - perm_Bk[[t]], "F")))
  }
  switch_of_pred <- P[which.min(norm_of_true_and_pred_Bk),]

  for(j in 2:4){
    pred[[j]] <- pred[[j]][switch_of_pred]
  }
 
  new_pred_Bk <- pred$Bk
  Number_of_nonzeros <- sapply(true_Bk, function(t) length(which(t != 0)))

  tpr <- vector(length = K)
  for(k in 1:K){
    selected <- (ifelse(abs(new_pred_Bk[[k]]) < 5e-6, 0, new_pred_Bk[[k]]) != 0)
    tpr[k] <- length(which(true_Bk[[k]] != 0 & selected)) / Number_of_nonzeros[k]
  }
  
  return(mean(tpr))
}

FPR <- function(true, pred){
  true_Bk <- true$Bk
  pred_Bk <- pred$Bk
  
  P <- permutations(K, K)
  norm_of_true_and_pred_Bk <- vector()
  for(i in 1:nrow(P)){
    perm_Bk <- pred_Bk[P[i,]]
    norm_of_true_and_pred_Bk[i] <- sum(sapply(1:K, function(t) norm(true_Bk[[t]] - perm_Bk[[t]], "F")))
  }
  switch_of_pred <- P[which.min(norm_of_true_and_pred_Bk),]
  
  for(j in 2:4){
    pred[[j]] <- pred[[j]][switch_of_pred]
  }
  
  new_pred_Bk <- pred$Bk
  Number_of_zeros <- sapply(true_Bk, function(t) length(which(t == 0)))
  
  fpr <- vector(length = K)
  for(k in 1:K){
    selected <- (ifelse(abs(new_pred_Bk[[k]]) < 5e-6, 0, new_pred_Bk[[k]]) != 0)
    fpr[k] <- length(which(true_Bk[[k]] == 0 & selected)) / Number_of_zeros[k]
  }
  
  return(mean(fpr))
}

FDR <- function(true, pred){
  true_Bk <- true$Bk
  pred_Bk <- pred$Bk
  
  P <- permutations(K, K)
  norm_of_true_and_pred_Bk <- vector()
  for(i in 1:nrow(P)){
    perm_Bk <- pred_Bk[P[i,]]
    norm_of_true_and_pred_Bk[i] <- sum(sapply(1:K, function(t) norm(true_Bk[[t]]-perm_Bk[[t]], "F")))
  }
  switch_of_pred <- P[which.min(norm_of_true_and_pred_Bk),]
  
  for(j in 2:4){
    pred[[j]] <- pred[[j]][switch_of_pred]
  }
  
  new_pred_Bk <- lapply(1:K, function(x) ifelse(abs(pred$Bk[[x]]) < 5e-6, 0, pred$Bk[[x]]))
  
  fdr <- vector(length = K)
  R <- sapply(new_pred_Bk, function(t) length(which(t != 0)))
  for(k in 1:K){
    selected <- (ifelse(abs(new_pred_Bk[[k]]) < 5e-6, 0, new_pred_Bk[[k]]) != 0)
    fdr[k] <- ifelse(R[k] == 0, 0, length(which(true_Bk[[k]] == 0 & selected)) / R[k])
  }
  
  return(mean(fdr))
}

TPR_no_penalty <- function(true, pred){
  true_Bk <- true$Bk
  pred_Bk <- pred$Bk
  
  P <- permutations(K, K)
  norm_of_true_and_pred_Bk <- vector()
  for(i in 1:nrow(P)){
    perm_Bk <- pred_Bk[P[i,]]
    norm_of_true_and_pred_Bk[i] <- sum(sapply(1:K, function(t) norm(true_Bk[[t]]-perm_Bk[[t]], "F")))
  }
  switch_of_pred <- P[which.min(norm_of_true_and_pred_Bk),]
  
  for(j in 1:3){
    pred[[j]] <- pred[[j]][switch_of_pred]
  }
  
  new_pred_Bk <- pred$Bk
  Number_of_nonzeros <- sapply(true_Bk, function(t) length(which(t != 0)))
  
  tpr <- vector(length = K)
  for(k in 1:K){
    selected <- (ifelse(abs(new_pred_Bk[[k]]) < 5e-6, 0, new_pred_Bk[[k]]) != 0)
    tpr[k] <- length(which(true_Bk[[k]] != 0 & selected)) / Number_of_nonzeros[k]
  }
  
  return(mean(tpr))
}


FPR_no_penalty <- function(true, pred){
  true_Bk <- true$Bk
  pred_Bk <- pred$Bk
  
  P <- permutations(K, K)
  norm_of_true_and_pred_Bk <- vector()
  for(i in 1:nrow(P)){
    perm_Bk <- pred_Bk[P[i,]]
    norm_of_true_and_pred_Bk[i] <- sum(sapply(1:K, function(t) norm(true_Bk[[t]]-perm_Bk[[t]], "F")))
  }
  switch_of_pred <- P[which.min(norm_of_true_and_pred_Bk),]
  
  for(j in 1:3){
    pred[[j]] <- pred[[j]][switch_of_pred]
  }
  
  new_pred_Bk <- pred$Bk
  Number_of_zeros <- sapply(true_Bk, function(t) length(which(t == 0)))
  
  fpr <- vector(length = K)
  for(k in 1:K){
    selected <- (ifelse(abs(new_pred_Bk[[k]]) < 5e-6, 0, new_pred_Bk[[k]]) != 0)
    fpr[k] <- length(which(true_Bk[[k]] == 0 & selected)) / Number_of_zeros[k]
  }
  
  return(mean(fpr))
}

FDR_no_penalty <- function(true, pred){
  true_Bk <- true$Bk
  pred_Bk <- pred$Bk
  
  P <- permutations(K, K)
  norm_of_true_and_pred_Bk <- vector()
  for(i in 1:nrow(P)){
    perm_Bk <- pred_Bk[P[i,]]
    norm_of_true_and_pred_Bk[i] <- sum(sapply(1:K, function(t) norm(true_Bk[[t]]-perm_Bk[[t]], "F")))
  }
  switch_of_pred <- P[which.min(norm_of_true_and_pred_Bk),]
  
  for(j in 1:3){
    pred[[j]] <- pred[[j]][switch_of_pred]
  }
  
  new_pred_Bk <- lapply(1:K, function(x) ifelse(abs(pred$Bk[[x]]) < 5e-6, 0, pred$Bk[[x]]))
  
  fdr <- vector(length = K)
  R <- sapply(new_pred_Bk, function(t) length(which(t != 0)))
  for(k in 1:K){
    selected <- (ifelse(abs(new_pred_Bk[[k]]) < 5e-6, 0, new_pred_Bk[[k]]) != 0)
    fdr[k] <- ifelse(R[k] == 0, 0, length(which(true_Bk[[k]] == 0 & selected)) / R[k])
  }
  
  return(mean(fdr))
}

MSE <- function(true, pred){
  true_pi <- true$pi
  true_Bk <- true$Bk
  true_sigma <- true$sigma
  pred_Bk <- pred$Bk
  
  P <- permutations(K, K)
  norm_of_true_and_pred_Bk <- vector()
  for(i in 1:nrow(P)){
    perm_Bk <- pred_Bk[P[i,]]
    norm_of_true_and_pred_Bk[i] <- sum(sapply(1:K, function(t) norm(true_Bk[[t]]-perm_Bk[[t]], "F")))
  }
  switch_of_pred <- P[which.min(norm_of_true_and_pred_Bk),]
  
  for(j in 2:4){
    pred[[j]] <- pred[[j]][switch_of_pred]
  }
  new_pred_pi <- pred$pi
  new_pred_Bk <- pred$Bk
  new_pred_sigma <- pred$sigma
  
  mse <- list(pi = vector(length = K), Bk = vector(length = K), sigma = vector(length = K))
  for(k in 1:K){
    mse$pi[k] <- norm(true_pi[k] - new_pred_pi[k], "2")^2
    mse$Bk[k] <- norm(true_Bk[[k]] - new_pred_Bk[[k]], "F")^2 / length(new_pred_Bk[[k]])
    mse$sigma[k] <- norm(true_sigma[[k]] - new_pred_sigma[[k]], "F")^2 / length(new_pred_sigma[[k]])
  }
  
  return(unlist(lapply(mse, mean)))
}


MSE_no_penalty <- function(true, pred){
  true_pi <- true$pi
  true_Bk <- true$Bk
  true_sigma <- true$sigma
  pred_Bk <- pred$Bk
  
  P <- permutations(K, K)
  norm_of_true_and_pred_Bk <- vector()
  for(i in 1:nrow(P)){
    perm_Bk <- pred_Bk[P[i,]]
    norm_of_true_and_pred_Bk[i] <- sum(sapply(1:K, function(t) norm(true_Bk[[t]] - perm_Bk[[t]], "F")))
  }
  switch_of_pred <- P[which.min(norm_of_true_and_pred_Bk),]
  
  for(j in 1:3){
    pred[[j]] <- pred[[j]][switch_of_pred]
  }
  new_pred_pi <- pred$pi
  new_pred_Bk <- pred$Bk
  new_pred_sigma <- pred$sigma
  
  mse <- list(pi = vector(length = K), Bk = vector(length = K), sigma = vector(length = K))
  for(k in 1:K){
    mse$pi[k] <- norm(true_pi[k] - new_pred_pi[k], "2")^2
    mse$Bk[k] <- norm(true_Bk[[k]] - new_pred_Bk[[k]], "F")^2 / length(new_pred_Bk[[k]])
    mse$sigma[k] <- norm(true_sigma[[k]] - new_pred_sigma[[k]], "F")^2 / length(new_pred_sigma[[k]])
  }
  
  return(unlist(lapply(mse, mean)))
}

floor_6 <- function(x) as.data.frame(trunc(x*10^5)/10^5)

pred_ll_dens <- function(dat, opt){
  true_B <- dat$true$Bk
  opt_B <- opt$Bk
  opt_sigma <- opt$sigma
  opt_inv_sigma <- lapply(opt$sigma, solve)
  test_x <- dat$test_X
  test_y <- dat$test_Y
  
  dens_list <- lapply(1:length(opt$Bk),
                      function(x) density_f(test_x, test_y,
                                            opt_B[[x]], opt_sigma[[x]], opt_inv_sigma[[x]]))
  return(dens_list)
}

predictive_ll <- function(opt, dens, l, n){
  
  pi_ <- opt[[l]]$pi
  dens <- dens[[l]]
  opt_K <- length(pi_)
  
  in_log <- vector(length = n)
  D <- do.call(cbind, dens)      
  in_log <- as.vector(D %*% pi_)
  
  pred_ll <- (-2)*(1/n)*sum(log(in_log))
  
  return(pred_ll)
}

File: Python/Scripts/Functions/simulation_summary_functions.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import itertools
import numpy as np
import pandas as pd


# In[2]:


def matrix_norm(x):
    x = np.asarray(x)

    if x.ndim >= 2:
        return np.linalg.norm(x, ord = "fro")
    else:
        return np.linalg.norm(x)


def find_best_permutation(true_Bk, pred_Bk):
    k = len(true_Bk)
    permutations = list(itertools.permutations(range(k)))

    norm_values = []
    for perm in permutations:
        total_norm = sum(
            matrix_norm(
                np.asarray(true_Bk[j]) -
                np.asarray(pred_Bk[perm[j]])
            )
            for j in range(k)
        )

        norm_values.append(total_norm)

    best_perm = permutations[np.argmin(norm_values)]

    return best_perm


def align_prediction(true, pred):
    true_Bk = true["Bk"]
    pred_Bk = pred["Bk"]

    best_perm = find_best_permutation(true_Bk, pred_Bk)

    aligned_pred = pred.copy()

    aligned_pred["Bk"] = [pred_Bk[i] for i in best_perm]

    if "pi" in pred:
        aligned_pred["pi"] = np.asarray(pred["pi"])[list(best_perm)]

    if "sigma" in pred:
        aligned_pred["sigma"] = [pred["sigma"][i] for i in best_perm]

    return aligned_pred


def threshold_prediction(Bk, threshold = 5e-6):

    return [
        np.where(
            np.abs(np.asarray(B)) < threshold,
            0,
            np.asarray(B)
        )
        for B in Bk
    ]


# In[3]:


def TPR(true, pred):
    true_Bk = true["Bk"]

    aligned_pred = align_prediction(true, pred)
    new_pred_Bk = aligned_pred["Bk"]

    new_pred_Bk = threshold_prediction(new_pred_Bk)

    number_of_nonzeros = [np.count_nonzero(B) for B in true_Bk]

    tpr = []
    for k in range(len(true_Bk)):
        true_nonzero = np.asarray(true_Bk[k]) != 0
        selected = np.asarray(new_pred_Bk[k]) != 0

        numerator = np.count_nonzero(true_nonzero & selected)

        denominator = number_of_nonzeros[k]

        tpr.append(np.divide(numerator, denominator))

    return np.mean(tpr)


# In[4]:


def FPR(true, pred):
    true_Bk = true["Bk"]

    aligned_pred = align_prediction(true, pred)
    new_pred_Bk = aligned_pred["Bk"]

    new_pred_Bk = threshold_prediction(new_pred_Bk)

    number_of_zeros = [np.count_nonzero(B == 0) for B in true_Bk]

    fpr = []
    for k in range(len(true_Bk)):
        true_zero = np.asarray(true_Bk[k]) == 0
        selected = np.asarray(new_pred_Bk[k]) != 0

        numerator = np.count_nonzero(true_zero & selected)

        denominator = number_of_zeros[k]

        fpr.append(np.divide(numerator, denominator))

    return np.mean(fpr)


# In[5]:


def FDR(true, pred):
    true_Bk = true["Bk"]

    aligned_pred = align_prediction(true, pred)
    new_pred_Bk = threshold_prediction(aligned_pred["Bk"])

    fdr = []
    R = [np.count_nonzero(B != 0) for B in new_pred_Bk]

    for k in range(len(true_Bk)):
        if R[k] == 0:
            fdr.append(0)
        else:
            true_zero = np.asarray(true_Bk[k]) == 0
            selected = np.asarray(new_pred_Bk[k]) != 0

            numerator = np.count_nonzero(true_zero & selected)

            fdr.append(numerator / R[k])

    return np.mean(fdr)


# In[6]:


def MSE(true, pred):
    true_pi = np.asarray(true["pi"])
    true_Bk = true["Bk"]
    true_sigma = true["sigma"]

    aligned_pred = align_prediction(true, pred)

    new_pred_pi = np.asarray(aligned_pred["pi"])
    new_pred_Bk = aligned_pred["Bk"]
    new_pred_sigma = aligned_pred["sigma"]

    mse_pi = []
    mse_Bk = []
    mse_sigma = []
    for k in range(len(true_Bk)):
        mse_pi.append((true_pi[k] - new_pred_pi[k]) ** 2)

        Bk_diff = (np.asarray(true_Bk[k]) - np.asarray(new_pred_Bk[k]))

        mse_Bk.append(matrix_norm(Bk_diff) ** 2 / np.asarray(new_pred_Bk[k]).size)

        sigma_diff = (np.asarray(true_sigma[k]) - np.asarray(new_pred_sigma[k]))

        mse_sigma.append(matrix_norm(sigma_diff) ** 2 / np.asarray(new_pred_sigma[k]).size)

    mse = {
        "pi": np.mean(mse_pi),
        "Bk": np.mean(mse_Bk),
        "sigma": np.mean(mse_sigma)
    }

    return np.array([
        mse["pi"],
        mse["Bk"],
        mse["sigma"]
    ])


# In[7]:


def floor_6(x):
    return np.trunc(x * 1e5) / 1e5

def pred_ll_dens(dat, opt):
    opt_B = opt["Bk"]
    opt_sigma = opt["sigma"]

    opt_inv_sigma = [np.linalg.inv(sigma) for sigma in opt_sigma]

    test_x = dat["test_X"]
    test_y = dat["test_Y"]

    dens_list = [density_f(test_x, test_y, opt_B[k], opt_sigma[k], opt_inv_sigma[k]) for k in range(len(opt_B))]

    return dens_list

def predictive_ll(opt, dens, l, n):
    pi_ = np.asarray(opt[l]["pi"])
    dens_l = dens[l]

    D = np.column_stack(dens_l)

    in_log = D @ pi_

    pred_ll = (-2 * (1/n) * np.sum(np.log(in_log)))

    return pred_ll

